### PR TITLE
Create pkg/config/model & pkg/config/env modules

### PR DESCRIPTION
**Pull request by dineshg13**
### What does this PR do?
For motivation refer [RFC](https://docs.google.com/document/d/1uKWa9rMNm0U-euEzEJ69KW7Id5u1E4GfYHyzC1SzYoI/edit#heading=h.wnkz5u7vx1fb)

This PR creates a `pkg/conf` package, a go module. Our eventual goal is to convert `comp/core/config` into a go module. We move essential parts from `pkg/config` to `pkg/conf` and create aliases. 
Please note:
1. We will use conf.Reader in modules that require config.Datadog and send the  `config.Datadog` instance as applicable.
2. Either config component or `config.Datadog` can be passed to modules after all the changes.

In the follow-up PRs, we will create three subfolders 
1. `pkg/conf/logssetup` - a folder for files essential for logs setup (moved from `pkg/config` ) 
2. `pkg/conf/load` is a folder for files responsible for loading config. 
3. `pkg/conf/setup` is the folder for initial setup files

### Reviewer's Checklist

- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [x] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [x] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

**Review by songy23 — APPROVED** [review]
LGTM from OTel end

**Review by gh123man — COMMENTED** [review]
The Code LGTM, However I feel that having `pkg/conf` and `pkg/config` living next to each other may be confusing, especially for folks that are not in the loop with why both have to exist. 

Would it be unreasonable to name `pkg/conf` something like `pkg/config_module`? (do others have opinions?)

**Comment by dineshg13** [comment]
I would prefer not to use the package name with a module. Ideally, after the refactor we are planning, every package/component can use `pkg/conf` or `comp/core/config`. Slowly, we should plan to deprecate `pkg/config.`

**Review by gh123man — APPROVED** [review]

**Review by pgimalac — COMMENTED** [review]
Can you add the modules you created to `tasks/modules.py` directly please ?

If you can't just export `pkg/config`, wouldn't it be simpler/cleaner to export submodules (ie. `pkg/config/env`, `pkg/config/logssetup`, etc) rather than create a `pkg/conf` ?

**Comment by dineshg13** [comment]
@pgimalac For my purpose or agent, it doesn't make a lot of difference whether we create `pkg/config/env` or `pkg/conf/env.` But if we want to use config external to Agent mono repo, having a single module with the types and initialization code makes more sense. 

I will add the module in `tasks/modules.py` .

File: pkg/conf/utils.go [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+package conf
+
+import "os"
+
+// PathExists returns true if the given path exists
+func PathExists(path string) bool {
+	_, err := os.Stat(path)
+	return !os.IsNotExist(err)
+}
```
Comment (hush-hush) [inline pkg/conf/utils.go:14]: This is already available in `pkg/util/filesystem`

File: pkg/util/system/socket.go [hunk 2]
```diff
@@ -8,48 +8,10 @@
 package system
```
Comment (hush-hush) [inline pkg/util/system/socket.go:8]: Why keep this file ? Can't we update the code using `CheckSocketAvailable` to use `pkg/util/system/socket` instead ?

File: pkg/util/system/socket_windows.go [hunk 3]
```diff
@@ -5,34 +5,8 @@
```
Comment (hush-hush) [inline pkg/util/system/socket_windows.go:5]: Same here.

**Review by hush-hush — CHANGES_REQUESTED** [review]
Having two package `pkg/config` and `pkg/conf` is very confusing. Could it be possible to create sub-package instead ? Something like `pkg/config/env` and `pkg/config/module` for example ?

File: pkg/config/model/config_overrides.go [hunk 4]
```diff
@@ -3,9 +3,12 @@
 // This product includes software developed at Datadog (https://www.datadoghq.com/).
 // Copyright 2016-present Datadog, Inc.
 
-package config
+package model
 
-var overrideVars = make(map[string]interface{})
+var (
+	overrideVars  = make(map[string]interface{})
+	overrideFuncs = make([]func(Config), 0)
```
Comment (pgimalac) [inline pkg/config/model/config_overrides.go:10]: Do you really want to redefine that here ? Or is it supposed to replace the definition from `pkg/config/config.go:110` (which you did not remove)

File: pkg/util/system/socket/go.mod [hunk 5]
```diff
@@ -0,0 +1,11 @@
+module github.com/DataDog/datadog-agent/pkg/util/system/socket
```
Comment (pgimalac) [inline pkg/util/system/socket/go.mod:1]: Wouldn't it be simpler to make a module out of `pkg/util/system` in case you need something else than `socket` ?

**Review by pgimalac — COMMENTED** [review]
Thanks a lot for the changes !
Just one thing I'm not sure about, and a suggestion

Can you add the new modules to `tasks/modules.py` in this PR ?

File: pkg/util/system/socket/socket.go [hunk 6]
```diff
@@ -5,7 +5,7 @@
 
 //go:build !windows
 
-package system
+package socket
```
Comment (hush-hush) [inline pkg/util/system/socket/socket.go:8]: Since the package name is now `socket` could we rename the function:
- `CheckSocketAvailable` → `CheckAvailable` (or maybe `IsAvailable`, or simply `Available` ?)
- `checkSocketExists` → `checkExists`

File: pkg/config/aliases.go [hunk 7]
```diff
@@ -0,0 +1,83 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package config
```
Comment (hush-hush) [inline pkg/config/aliases.go:6]: I'm not a fan of duplicating our globals/constants to avoid editing the code using `config.XXX`. Updating the code base would produce a much bigger PR but would improve the final result in my opinion.

We could do this in another PRs if it's easier.

File: pkg/config/model/config_overrides.go (hunk 4 above)
Comment (hush-hush) [reply pkg/config/model/config_overrides.go:10]: I think we need this here, but should remove the one in `pkg/config/config.go`

**Review by hush-hush — CHANGES_REQUESTED** [review]

File: pkg/config/aliases.go (hunk 7 above)
Comment (pgimalac) [reply pkg/config/aliases.go:6]: I agree, I think every time I've seen this kind of change we just edited the calls all over the repo to keep a clean state (ie. avoid having re-exports of a module inside another one)

Comment (hush-hush) [reply pkg/config/aliases.go:6]: Perfect, could you create a card in your backlog so we can track this ?

File: pkg/util/system/socket/go.mod (hunk 5 above)
Comment (dineshg13) [reply pkg/util/system/socket/go.mod:1]: `pkg/util/system` has dependency on `pkg/util/winutil`, that internally has other dependencies. Hence this approach.

Comment (pgimalac) [reply pkg/util/system/socket/go.mod:1]: That's fair

**Review by JLineaweaver — APPROVED** [review]

File: pkg/config/model/config_overrides.go (hunk 4 above)
Comment (dineshg13) [reply pkg/config/model/config_overrides.go:10]: done. removed one in config.go

File: pkg/config/aliases.go (hunk 7 above)
Comment (dineshg13) [reply pkg/config/aliases.go:6]: Created https://datadoghq.atlassian.net/browse/OTEL-1134 to track this

**Review by hush-hush — APPROVED** [review]

**Review by chouetz — APPROVED** [review]
OK for agent-platform files

File: pkg/config/model/config_overrides.go [hunk 8]
```diff
@@ -3,9 +3,12 @@
 // This product includes software developed at Datadog (https://www.datadoghq.com/).
 // Copyright 2016-present Datadog, Inc.
 
-package config
+package model
```
Comment (vboulineau) [inline pkg/config/model/config_overrides.go:6]: The name `model` makes sense when you are seeing the folder structure, e.g. `pkg/config/model`.
However when used in the code you end up with `model.XXX`, which does not make any sense to the reader.

Given that we're going to remove the aliases, I hope very soon, I think it's important to find a better name.

**Review by vboulineau — COMMENTED** [review]
Generally speaking LGTM, see comments

File: pkg/util/system/socket/socket.go [hunk 9]
```diff
@@ -14,11 +15,11 @@ import (
 	"time"
 )
 
-// CheckSocketAvailable returns if a socket at path is available
+// IsAvailable returns if a socket at path is available
 // first boolean returns if socket path exists
 // second boolean returns if socket is reachable
-func CheckSocketAvailable(path string, timeout time.Duration) (bool, bool) {
-	if !checkSocketExists(path) {
+func IsAvailable(path string, timeout time.Duration) (bool, bool) {
```
[the remark this answers is not in the document]
Comment (dineshg13) [reply pkg/util/system/socket/socket.go:21]: The package is `socket`. So it would be `socket.IsAvailable`

File: pkg/config/model/config_overrides.go (hunk 8 above)
Comment (dineshg13) [reply pkg/config/model/config_overrides.go:6]: We iterated on name few times. Initial proposal was `pkg/conf`. We settled on `pkg/config/model` as middle ground. Do you any suggestions for name ?

File: pip.conf
[the remark this answers is not in the document]
Comment (dineshg13) [reply pip.conf:1]: Not sure how it got added. Let me remove it

File: pkg/config/model/config_overrides.go (hunk 8 above)
Comment (vboulineau) [reply pkg/config/model/config_overrides.go:6]: We can do `pkg/config/config` as we're going to remove everything in `pkg/config` anyway.

File: pkg/config/env/environment_detection.go [hunk 10]
```diff
@@ -74,28 +75,35 @@ func IsFeaturePresent(feature Feature) bool {
...
 // IsAutoconfigEnabled returns if autoconfig from environment is activated or not
-func IsAutoconfigEnabled() bool {
+func IsAutoconfigEnabled(cfg model.Reader) bool {
 	// Usage of pure environment variables should be deprecated
 	for _, envVar := range []string{autoconfEnvironmentVariable, autoconfEnvironmentVariableWithTypo} {
 		if autoconfStr, found := os.LookupEnv(envVar); found {
 			activateAutoconfFromEnv, err := strconv.ParseBool(autoconfStr)
 			if err != nil {
-				log.Errorf("Unable to parse Autoconf value: '%s', err: %v - autoconfig from environment will be deactivated", autoconfStr, err)
+				log.Errorf(
```
Comment (vboulineau) [inline pkg/config/env/environment_detection.go:84]: nit: seems like auto-formatting from IDE or something, please avoid it (and multi-line is IMHO way worse to read in these cases)

File: pkg/config/model/config_overrides.go (hunk 8 above)
Comment (dineshg13) [reply pkg/config/model/config_overrides.go:6]: I think `pkg/config/config` will confuse the developer because of aliases; both types would be present in two packages with the same name.  I don't have clear picture what would be the final state of `pkg/config` . We can revisit this in future .

Comment (dineshg13) [reply pkg/config/model/config_overrides.go:6]: Discussed offline. 
1. We will be using import aliases like  `configModel "pkg/config/model" ` to denote that it is config model . 
2. We will minimize the code that refers `pkg/config/model` and try to use config component throughout the code. 
We will use pkg/config/model in 4 places 
1.  To init config component 
2. To create a mock config for tests 
3. In the DD exporter.
4. Non-comp code that needs to be converted to go module would use pkg/config/model
